_posts/WAI-is-the-answer.md

```diff
@@ -0,0 +1,569 @@
+---
+title: 'WAI is the Answer !!!'
+excerpt: 'Porting libraries to npm and pip made easy with WAI'
+date: '2023-02-08T13:30:37.390Z'
+author: Rudra
+published: true
+---
+
+## _Ding Dong_
+
+“**Sir, we are here to move your package.**”
+
+![https://www.movingsolutions.in/blog/wp-content/uploads/2021/02/iba-approved-packers-and-movers-in-ghaziabad-and-noida.jpg](https://www.movingsolutions.in/blog/wp-content/uploads/2021/02/iba-approved-packers-and-movers-in-ghaziabad-and-noida.jpg)
+
+**That’s right** we at Wasmer move your packages for free
+
+> 🗒️ Note: Package refers to a library/executable that you might have developed for a certain language
+
+# What are we going to do?
+
+For this tutorial we will be publishing [the `sgp4` crate](https://github.com/neuromorphicsystems/sgp4/), a library used to predict the location of satellites in orbit after a certain amount of time.
+
+You can find the [dynamite-bud/sgp4](https://wapm.io/dynamite-bud/sgp4) package on WAPM.
+
+At the end we’ll be able to use `sgp4` from JavaScript like this.
+
+```javascript
+const resolveResult = ({ tag, val }) => (tag === 'err' ? new Error(val) : val);
+
+const wasm = await bindings.sgp4();
+let response = await fetch(
+  'https://celestrak.com/NORAD/elements/gp.php?GROUP=galileo&FORMAT=json',
+);
+let elementsArr = (await response.json()).map((e) =>
+  resolveResult(Elements.fromJson(wasm, JSON.stringify(e))),
+);
+for (let elements of elementsArr) {
+  console.log(elements.getObjectName());
+  let constants = resolveResult(Constants.fromElements(wasm, elements));
+  for (let hours of [12, 24]) {
+    console.log(`    t = ${hours * 60} min`);
+    let prediction = resolveResult(constants.propagate(parseFloat(hours * 60)));
+    console.log(`        r = ${prediction.position} km`);
+    console.log(`        ṙ = ${prediction.velocity} km.s⁻¹`);
+  }
+}
+```
+
+Here we used the official introductory example from original `sgp4`.
+
+## What is WAI ?
+
+[WAI](https://github.com/wasmerio/wai) or WebAssembly Interfaces is an _“[Interface Definition Language](https://en.wikipedia.org/wiki/Interface_description_language) (IDL)”_ like Protobufs or WebIDL, for interaction with multiple languages.
+
+Files for WAI use the `*.wai` extension.
+
+The WAI project has a variety of code generators, however the one we’ll be using today is `wai_bindgen_rust`. This Rust crate uses macros to take a WAI file and generate all the glue code we need to implement a WebAssembly module in Rust.
+
+## Let’s Port!
+
```

Comment:
I'd add a quick summary on what we are going to see below, so the reader can see how far they are in the process

1. Installation
2. Setup the enviroment
3. Create and use the WAI
4. Publish to WAPM

Otherwise it's a bit